editor/src/messages/tool/transform_layer/transform_layer_message_handler.rs

```diff
@@ -537,10 +565,17 @@ impl MessageHandler<TransformLayerMessage, TransformLayerMessageContext<'_>> for
 					match self.transform_operation {
 						TransformOperation::None => {}
 						TransformOperation::Grabbing(translation) => {
-							let delta_pos = input.mouse.position - self.mouse_position;
-							let delta_pos = (self.initial_transform * document_to_viewport.inverse()).transform_vector2(delta_pos);
-							let delta_viewport = if self.slow { delta_pos / SLOWING_DIVISOR } else { delta_pos };
-							let delta_scaled = delta_viewport / document_to_viewport.y_axis.length(); // Values are local to the viewport but scaled so values are relative to the current scale.
+							let viewport_delta = input.mouse.position - self.mouse_position;
+							let document_space_transform = self.initial_transform * document_to_viewport.inverse();
+							let delta_pos = if self.state.is_transforming_in_local_space {
+								let [local_x_axis, local_y_axis] = self.state.local_transform_axes;
+								let local_components = DVec2::new(viewport_delta.dot(local_x_axis), viewport_delta.dot(local_y_axis));
+								document_space_transform.matrix2 * local_components
+							} else {
+								document_space_transform.transform_vector2(viewport_delta)
+							};
+							let scale = document_to_viewport.y_axis.length();
+							let delta_scaled = (if self.slow { delta_pos / SLOWING_DIVISOR } else { delta_pos }) / scale;
```

Comment:
Along with adding significant complexity here, you have also broken the «local x-axis constraint» feature (press <kbd>G</kbd> to grab then press <kbd>X</kbd> twice.

https://github.com/user-attachments/assets/2a427529-5721-4cce-b9cc-0ccff8896b34

I don't see why there should be any changes to the functionality of the system, since the issue concerned only changing the overlays.